### `Problem 1`: Write a program that will give you in hand monthly salary after deduction on CTC - HRA(10%), DA(5%), PF(3%) and taxes deduction as below:

> Salary(Lakhs) : Tax(%)

*   Below 5 : 0%
*   5-10 : 10%
*   10-20 : 20%
*   aboove 20 : 30%

In [2]:
from datetime import datetime as dt

In [ ]:
# User input
salary = int(input('Enter your Monthly Salary :'))  


print(f'Your Salary is {salary} Lakhs\n')

# Tax deduction 
if salary < 5:
    tax_rate = 0 
elif 5 <= salary < 10:
    tax_rate = 0.10 
elif 10 <= salary < 20:
    tax_rate = 0.20  
else:
    tax_rate = 0.30  #


salary -= salary * tax_rate
print(f'After tax deduction, your salary is {salary:.2f} Lakhs\n')

# Common deductions: HRA(10%), DA(5%), PF(3%) → Total 18%
salary -= salary * 0.18
print(f'After all deductions, your Monthly salary is {salary:.2f} Lakhs\n')

In [35]:
# User input
salary = int(input('Enter your Monthly Salary :'))  


print(f'Your Salary is {salary} Lakhs\n')

# Tax deduction 
if salary < 5:
    print(f'You fall in (0-5) -> 0% Tax Brackets.\n')
    tax_rate = 0 
elif 5 <= salary < 10:
    print(f'You fall in (5-10)) -> 10% Tax Brackets.\n')
    tax_rate = 0.10 
elif 10 <= salary < 20:
    print(f'You fall in (10-20) -> 20% Tax Brackets.\n')
    tax_rate = 0.20  
else:
    print(f'You fall in (above 20) 30% Tax Brackets.\n')
    tax_rate = 0.30  


salary -= salary * tax_rate
print(f'After tax deduction, your salary is {salary:.2f} Lakhs\n')

# Common deductions: HRA(10%), DA(5%), PF(3%) → Total 18%
salary -= salary * 0.18
print(f'After all deductions, your Monthly salary is {salary:.2f} Lakhs\n')

Your Salary is 25 Lakhs

You fall in (above 20) 30% Tax Brackets.

After tax deduction, your salary is 17.50 Lakhs

After all deductions, your Monthly salary is 14.35 Lakhs



### `Problem 2`: Write a program that take a user input of three angles and will find out whether it can form a triangle or not.

In [ ]:
# Write code here

# user input
a = int(input("Enter 'A' Angle :"))
b = int(input("Enter 'B' Angle :"))
c = int(input("Enter 'C' Angle :"))
print(f'Your Angles are {a}, {b}, {c}')

# triangle form logic
if a + b + c == 180:
    print('This is a Triangle \n')
else:
    print('This is not a Triangle.')

Your Angles are 60, 45, 90
This is not a Triangle.


### `Problem 3`: Write a program that will take user input of cost price and selling price and determines whether its a loss or a profit.

In [ ]:
# user input
selling_price = int(input('Enter the selling price :'))
cost_price = int(input('Enter the cost price :'))
print(f'your selling price is {selling_price}, cost price is {cost_price} \n')
# condition 
if selling_price > cost_price:
    print('Your are in profit !')
else:
    print('Your are in the Loss !')

your selling price is 20, cost price is 256 

Your are in the Loss


### `Problem 4`: Write a menu-driven program -
1. cm to ft
2. km to miles
3. USD to INR
4. exit

In [ ]:
# Write code here

print(""""
      Select the option which function want to active
      
      1. cm to ft
      2. km to miles
      3. USD to INR
      4. exit
      
      """)

option = int(input('Enter the digit from 1 to 4 \n'))

# Check the user choose the valid options
if option < 5:
    # Option Working 
    if option == 1:
        cm = int(input('Enter the cm : \n'))
        print(f'Your cm is {cm} and That is {cm * 0.0328084} ft \n')
    elif option == 2:
        km = int(input('Enter the km : \n'))
        print(f'Your cm is {km} and That is {km * 0.621371} miles \n')
    elif option == 2:
        usd = int(input('Enter the km : \n'))
        print(f'Your USD is {usd} and That is {usd * 85.66} INR \n')
    else:
        print('Thank you ! Visit Our service again .....')
else:
    print('Select the Correct Option !')

"
      Select the option which function want to active
      
      1. cm to ft
      2. km to miles
      3. USD to INR
      4. exit
      
      
Your cm is 2535 and That is 83.16929400000001 ft 



### `Problem 5` - Exercise 12: Display Fibonacci series up to 10 terms.

*Note: The Fibonacci Sequence is a series of numbers. The next number is found by adding up the two numbers before it. The first two numbers are 0 and 1. For example, 0, 1, 1, 2, 3, 5, 8, 13, 21. The next number in this series above is 13+21 = 34*

<img src='https://media.geeksforgeeks.org/wp-content/uploads/20230828183030/FIBONACCI-SERIES.png' width=500, height=500>

[How the code is work](https://pythontutor.com/render.html#mode=display)

In [71]:
# Write code here

n1, n2 = 0, 1

for i in range(11):
    print(n1)
    
    next = n1 + n2
    
    n1 = n2 
    n2 = next

0
1
1
2
3
5
8
13
21
34
55


### `Problem 6` - Find the factorial of a given number.

Write a program to use the loop to find the factorial of a given number.

The factorial (symbol: `!`) means to multiply all whole numbers from the chosen number down to 1.

For example: calculate the factorial of 5

```bash
5! = 5 × 4 × 3 × 2 × 1 = 120
```

Output:

```bash
120
```

In [12]:
num = int(input('Enter a number :'))
print(f'Your number is {num}')
fact = 1

for i in range(1, num+1):
    fact = i*fact
    
print('your factorial is :', fact)

Your number is 5
your factorial is : 120


### `Problem 7` - Reverse a given integer number.

Example:

`Input:`

```bash
76542
```

`Output:`

```bash
24567
```

In [20]:
num = int(input('Enter a integer :'))
print('Your Integer is :', num)

rev = 0

while num > 0:
    last = num % 10
    rev = rev * 10 + last
    num = num // 10

print('Your reverse number is :', rev)

Your Integer is : 13238
Your reverse number is : 83231


In [ ]:
# small example what happing in the code
n = 123

l = n % 10 
l

3

In [35]:
r = 0 * 10 + l
r

3

In [36]:
n // 10

12

### `Problem 8`: Take a user input as integer N. Find out the sum from 1 to N. If any number if divisible by 5, then skip that number. And if the sum is greater than 300, don't need to calculate the sum further more. Print the final result. And don't use for loop to solve this problem.

**Example 1:**

`Input:`

```bash
30
```

`Output:`

```bash
276
```

Break down 
- user input
- sum 1 to n

- n // 5 then skip 
- sum is > 300 stop

In [76]:
n = 30
print(f'Your given number is {n}\n')

sum = 0

for i in range(1, n+1):
    sum = i + sum 
    if i == 5:
        print('Which number I skipped :', i)
        continue 
    elif sum >= 300:
        break
    
    print(sum)

print(f'your total is {sum} \n')

Your given number is 30

1
3
6
10
Which number I skipped : 5
21
28
36
45
55
66
78
91
105
120
136
153
171
190
210
231
253
276
your total is 300 



### `Problem 9`: Write a program that keeps on accepting a number from the user until the user enters Zero. Display the sum and average of all the numbers.

In [96]:
# Write code here

a = int(input('Enter the number :'))
print(f'Your First number is {a}\n')

sum = a
avg = 0
count = 1 # already taken one from top

while True:
    a = int(input('Enter the number :'))
    print(f'Your {count + 1} number is {a}\n')
    
    if a == 0:
        print(f'Your {count}th is ZERO  so Stop  ... \n')
        break
    
    count += 1
    
    sum += a
    avg = sum / count
    

print(f'Your Average is : {avg} \n')
print(f'Your Sum is {sum} \n')

Your First number is 1

Your 2 number is 2

Your 3 number is 3

Your 4 number is 0

Your 3th is ZERO  so Stop  ... 

Your Average is : 2.0 

Your Sum is 6 



### `Problem 9`: Write a program which will find all such numbers which are divisible by 7 but are not a multiple of 5, between 2000 and 3200 (both included). The numbers obtained should be printed in a comma-separated sequence on a single line.

> *Breakdown the Problem*
- 2000 to 3200 both included
-  divisible by 7
- multiple of 5
- comma separated 

In [15]:
l = [ ]
for i in range(2000, 3201):
    if i % 7 == 0 and i % 5 != 0:
        l.append(str(i))
        
print(",".join(l))

2002,2009,2016,2023,2037,2044,2051,2058,2072,2079,2086,2093,2107,2114,2121,2128,2142,2149,2156,2163,2177,2184,2191,2198,2212,2219,2226,2233,2247,2254,2261,2268,2282,2289,2296,2303,2317,2324,2331,2338,2352,2359,2366,2373,2387,2394,2401,2408,2422,2429,2436,2443,2457,2464,2471,2478,2492,2499,2506,2513,2527,2534,2541,2548,2562,2569,2576,2583,2597,2604,2611,2618,2632,2639,2646,2653,2667,2674,2681,2688,2702,2709,2716,2723,2737,2744,2751,2758,2772,2779,2786,2793,2807,2814,2821,2828,2842,2849,2856,2863,2877,2884,2891,2898,2912,2919,2926,2933,2947,2954,2961,2968,2982,2989,2996,3003,3017,3024,3031,3038,3052,3059,3066,3073,3087,3094,3101,3108,3122,3129,3136,3143,3157,3164,3171,3178,3192,3199


### `Problem 10`: Write a program, which will find all such numbers between 1000 and 3000 (both included) such that each digit of the number is an even number. The numbers obtained should be printed in a space-separated sequence on a single line.



> Break 

- 1000 to 3000 both included
- even num (total num is divisible by 2 )
- even digit (last int is divisible by 2)
- space separated in single line

In [26]:

l = []
last = [2, 4, 6, 8]

for i in range (1000, 3001):
    if i % 2 == 0:
        l.append(str(i))

print(','.join(l))

1000,1002,1004,1006,1008,1010,1012,1014,1016,1018,1020,1022,1024,1026,1028,1030,1032,1034,1036,1038,1040,1042,1044,1046,1048,1050,1052,1054,1056,1058,1060,1062,1064,1066,1068,1070,1072,1074,1076,1078,1080,1082,1084,1086,1088,1090,1092,1094,1096,1098,1100,1102,1104,1106,1108,1110,1112,1114,1116,1118,1120,1122,1124,1126,1128,1130,1132,1134,1136,1138,1140,1142,1144,1146,1148,1150,1152,1154,1156,1158,1160,1162,1164,1166,1168,1170,1172,1174,1176,1178,1180,1182,1184,1186,1188,1190,1192,1194,1196,1198,1200,1202,1204,1206,1208,1210,1212,1214,1216,1218,1220,1222,1224,1226,1228,1230,1232,1234,1236,1238,1240,1242,1244,1246,1248,1250,1252,1254,1256,1258,1260,1262,1264,1266,1268,1270,1272,1274,1276,1278,1280,1282,1284,1286,1288,1290,1292,1294,1296,1298,1300,1302,1304,1306,1308,1310,1312,1314,1316,1318,1320,1322,1324,1326,1328,1330,1332,1334,1336,1338,1340,1342,1344,1346,1348,1350,1352,1354,1356,1358,1360,1362,1364,1366,1368,1370,1372,1374,1376,1378,1380,1382,1384,1386,1388,1390,1392,1394,1396,1398,

### `Problem 11`: A robot moves in a plane starting from the original point (0,0). The robot can move toward UP, DOWN, LEFT and RIGHT with a given steps.
The trace of robot movement is shown as the following:
```
UP 5
DOWN 3
LEFT 3
RIGHT 2
!
```
> The numbers after the direction are steps.

> `!` means robot stop there.

**Please write a program to compute the distance from current position after a sequence of movement and original point.**

*If the distance is a float, then just print the nearest integer.*

Example:

`Input`:
```
UP 5
DOWN 3
LEFT 3
RIGHT 2
!
```
`Output`:
```
2
```

In [ ]:
dt.now()

datetime.datetime(2025, 3, 28, 7, 35, 54, 270919)

| **Direction** | **Effect on (x, y)** |
|--------------|---------------------|
| UP `n`      | `y += n` (Increase y) |
| DOWN `n`    | `y -= n` (Decrease y) |
| LEFT `n`    | `x -= n` (Decrease x) |
| RIGHT `n`   | `x += n` (Increase x) |

In [ ]:
import math 

x = 0
y = 0

print("""
      
      Give steps to robot 
      
      left or l = left
      right or r = right 
      down or  d = down
      up or u = up 
      
      Type '!' fpr STOP 
      
      """)

while True:
    steps = input('Enter both : \n')
    
    
    if steps.lower() == '!':
        break
    
    direction = steps.lower().split()[0]
    steps = int(steps.split()[1])
    
    
    if direction == 'u' or  direction ==  'UP':
        y += steps
        print(f'Robot moved UP by {steps} steps. Now at ({x}, {y}).\n')
    elif direction == 'd' or  direction ==  'DOWN':
        y -= steps
        print(f'Robot move DOWN by {steps}. Now at({x}, {y}). \n')
    elif direction == 'l' or  direction ==  'LEFT':
        x -= steps
        print(f'Robot moved LEFT by {steps} steps. Now at ({x}, {y}).\n')
    elif direction == 'r' or  direction == 'RIGHT':
        x +=  steps
        print(f'Robot moved RIGHT by {steps} steps. Now at ({x}, {y}).\n')
    
distance = math.sqrt( (x**2) + (y**2) )
print(f'Euclidean distance : {round(distance)}. \n')


      Give steps to robot 
      
      left or l = left
      right or r = right 
      down or  d = down
      up or u = up 
      
      
Robot moved LEFT by 2 steps. Now at (-2, 0).

Robot moved RIGHT by 2 steps. Now at (0, 0).

Robot move DOWN by 2. Now at(0, -2). 

Robot moved UP by 2 steps. Now at (0, 0).

Euclidean distance : 0. 



In [16]:
dt.now()

datetime.datetime(2025, 3, 28, 8, 38, 42, 709950)

### `Problem 12`:Write a program to print whether a given number is a prime number or not

In [3]:
dt.now()

datetime.datetime(2025, 3, 28, 22, 50, 53, 250831)

In [12]:
num = int(input('Enter a number :'))

flag = True

for i in range(2, num):
    if num % i == 0:
        flag = False
        break

if flag == True:
    print(f'Yor given number {num} is Prime number.')
else:
    print(f'Yor given number {num} is not a Prime number.')

Yor given number 21 is not a Prime number.


### `Problem 13`:Print all the Armstrong numbers in a given range.
Range will be provided by the user<br>
Armstrong number is a number that is equal to the sum of cubes of its digits. For example 0, 1, 153, 370, 371 and 407 are the Armstrong numbers.

<img src='https://d1m75rqqgidzqn.cloudfront.net/wp-data/2020/02/14183658/Feb-14-blog-thumbnail-armstrong-number-1.png'  width='300' height='200'>

In [14]:
dt.now()

datetime.datetime(2025, 3, 28, 23, 1, 6, 141985)

In [26]:
start = int(input("Enter the start of the range: "))
end = int(input("Enter the end of the range: "))
print(f'Your Range is {start} to {end} \n ')


for num in range(start, end+1):
    temp = num
    sum = 0
    order = len(str(num))
    while temp > 0:
        digit = temp % 10
        sum += digit ** order
        temp //= 10
    if num == sum:
        print('That number is ', num)
        



Your Range is 0-1236 
 
That number is  0
That number is  1
That number is  2
That number is  3
That number is  4
That number is  5
That number is  6
That number is  7
That number is  8
That number is  9
That number is  153
That number is  370
That number is  371
That number is  407


### `Problem 14`:Calculate the angle between the hour hand and minute hand.

Note: There can be two angles between hands; we need to print a minimum of two. Also, we need to print the floor of the final result angle. For example, if the final angle is 10.61, we need to print 10.

Input:<br>
H = 9 , M = 0<br>
Output:<br>
90<br>
Explanation:<br>
The minimum angle between hour and minute
hand when the time is 9 is 90 degress.

In [17]:
(360/(12*60))

0.5

📌 Example Calculation for H = 9, M = 0
- **Hour Hand Angle**:  
 
  (9 \ 30) + (0 \ 0.5) = 270°

- **Minute Hand Angle**:  

  times 6 = 0°

- **Difference**:  

  |270 - 0| = 270°

- **Minimum Angle**:  

  min(270, 360 - 270) = 90°

- **Final Output**:  

  90


In [26]:
# input
h= int(input('Enter the Hour hand \n'))
m = int(input('Enter the Minute hand \n')) 
print(f'You asked the angle between {h} hr and {m} min \n ')

# validate the number is within the interval
if (h < 0 or m < 0 or h > 12 or m > 60):
    print('Wrong Input ... Try is Again . . ')

m_angle = m * (360/60)

h_angle = h * (360/12) + m * (360/(12*60))

angle = abs(h_angle - m_angle)

if angle> 180:
    print('Your angle is ', math.floor(360-angle))
else:
    print('Your angle is ', math.floor(angle))

You asked the angle between 9 hr and 0 min 
 
Your angle is  90


In [27]:
dt.now()

datetime.datetime(2025, 3, 28, 22, 3, 20, 632815)

`Explaination:` This code calculates the angle between the hour and minute hand on a clock. It takes user input for the hour and minute hand positions, validates the input to make sure it is within the range of a clock (0-12 for hours and 0-60 for minutes), then calculates the angles of the hour and minute hand based on their position on the clock. It then finds the difference between the two angles and takes the absolute value of that difference. It then checks if the angle is greater than 180, if it is, it prints the difference of 360 and the angle, else it prints the angle. The final output is the angle between the hands on the clock.

---

### `Problem 15`:Given two rectangles, find if the given two rectangles overlap or not. A rectangle is denoted by providing the x and y coordinates of two points: the left top corner and the right bottom corner of the rectangle. Two rectangles sharing a side are considered overlapping. (L1 and R1 are the extreme points of the first rectangle and L2 and R2 are the extreme points of the second rectangle).

Note: It may be assumed that the rectangles are parallel to the coordinate axis.

<img src='https://www.geeksforgeeks.org/wp-content/uploads/rectanglesOverlap.png' width='300' height='200'>

In [21]:
# Write code here
print("Enter both Rectangel points coordinates <x y> \n")

l1_x, l1_y = list(map(int, input('Enter L1 coordinate value  (Two value space seperated):').split()))
r1_x, r1_y = list(map(int, input('Enter r1 coordinate value,(Two value space seperated):').split()))
print(f'L1 coordinate point is <{l1_x}, {l1_y}> \nR1 coordinate point is <{r1_x}, {r1_y}> \n ')

l2_x, l2_y = list(map(int, input('Enter L2 coordinate value,(Two value space seperated):').split()))
r2_x, r2_y = list(map(int, input('Enter r2 coordinate value,(Two value space seperated):').split()))
print(f'L2 coordinate point is <{l2_x}, {l2_y}> \nR2 coordinate point is <{r2_x}, {r2_y}> \n')

# if rectangle has area 0, no overlap
if l1_x == r1_x or l1_y == r1_y or r2_x == l2_x or l2_y == r2_y:
  print("Don't Overlap")

# If one rectangle is on left side of other
elif l1_x > r2_x or l2_x > r1_x:
  print("Don't Overlap")

# If one rectangle is above other
elif r1_y > l2_y or r2_y > l1_y:
  print("Don't Overlap")
else:
  print("Overlap")

Enter both Rectangel points coordinates <x y>
L1 coordinate point is <2, 2> 
R1 coordinate point is <3, 3> 
 
L2 coordinate point is <4, 4> 
R2 coordinate point is <5, 5> 

Don't Overlap
